### PR TITLE
ARROW-12289: [C++] Create basic AsyncScanner implementation

### PR DESCRIPTION
**Pull request by westonpace**
Adds a naive implementation of `AsyncScanner` which is different from `SyncScanner` in a few ways:
* It does not use `ScanTask` and instead relies on `Fragment::ScanBatchesAsync` which returns `RecordBatchGenerator`.
* It does an unordered scan by default (i.e. batches from file N may arrive before all batches from file N-1 have arrived) and can order it if asked for
* It uses the unordered scan for `ToTable`.

It is "naive" because this PR does not add a complete implementation for `FileFragment::ScanBatchesAsync`.  This method relies on `FileFormat::ScanBatchesAsync` (in the same way that `FileFragment::Scan` relies on `FileFormat::ScanFile`).  This method (`FileFormat::ScanBatchesAsync`) _should_ be overridden in each of the formats (to rely on an async reader) but it is not (yet).

As a result, the performance for `AsyncScanner` is poor since it does not do any "per-file" parallelism nor does it do any "per-batch" parallelism.  Follow-up tasks are ARROW-12355 (CSV), ARROW-11772 (IPC), ARROW-11843 (Parquet)

In addition, this PR is built on top of ARROW-12287 so that will need to be merged first.  It will also need to rebase changes from ARROW-12161 and ARROW-11797.

### CODE REVIEW

File: cpp/src/arrow/dataset/scanner_test.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner_test.cc:49]: Ah, yes.  I was planning on adding whether to scan with `Scan` (to ensure we still test the legacy), `ScanBatches`, or `ScanBatchesUnordered` as a parameter as well.

Comment (westonpace) [reply cpp/src/arrow/dataset/scanner_test.cc:49]: I'll tackle that when I rebase ARROW-11797

File: cpp/src/arrow/dataset/dataset.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/dataset.cc]: Yep.  This logic was all backwards.  I've since changed it to a while loop.  Just pushed the change.

File: cpp/src/arrow/dataset/file_base.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc:144]: There is a silent precondition here that every fragment scan should return scan tasks that return at least 1 record batch (unless the entire fragment is empty in which case either 0 scan tasks or 1 scan task with 0 batches should both be ok).

I'm know this precondition holds for IPC and CSV (by virtue of there being only one scan task) but wasn't sure about parquet (i.e. can a push down filter cause a batch-less scan task to be emitted in the middle of a set of scan tasks?)

File: cpp/src/arrow/dataset/scanner.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.cc:490]: It will need to be.  The problem is that `MakeMergedGenerator` is immediately consuming `EnumeratingGenerator` which is not async-reentrant.  `MakeMergedGenerator` (erroneously) pulls from the outer (the gen_gen) generator in an async-reentrant fashion.  I'll make a follow-up JIRA just to keep this one simple.

Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.cc:490]: ARROW-12386

File: cpp/src/arrow/dataset/file_base.cc
Comment (lidavidm) [reply cpp/src/arrow/dataset/file_base.cc:144]: I may be reading this wrong, but when we finish one scan task and move on to the next, since we just fall through here, we'll return a completed Future which contains a nullptr, which gets returned as the result of `operator()`. So the generator's consumer will think that the generator has ended, even though we still have more scan tasks.

Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc:144]: Oh.  I think you are right.  I should probably add a scanner unit test that generates more than one scan task.  I'll work on that.

Comment (lidavidm) [reply cpp/src/arrow/dataset/file_base.cc:144]: I see that there's now a parameter to generate multiple scan tasks per fragment in InMemoryDataset - however, is that necessary? For one, it doesn't affect this code path, since this only affects file fragments. For another, it doesn't affect the scanner, which doesn't use scan tasks (directly); it'll use ScanBatchesAsync on the Fragment, which flattens all the scan tasks itself anyways.

So I think the issue pointed out here doesn't show up in test purely because only Parquet fragments expose multiple scan tasks per fragment right now.

Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc:144]: Fair point.  I think it's still necessary but could be renamed as it is a bit vague.  In the async case it is "batches per fragment" and in the sync case it is "scan tasks per fragment".  It was enough to break the async scanner (it currently fails these tests).  I also agree it doesn't expose this issue so I'll some more tests.

Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc:144]: Well, technically speaking the # of batches per fragment depends both on this and max batch size.  So I suppose we could have gotten sufficient testing by setting the max batch size small enough.  At the moment these tests help exercise the SyncScanner if nothing else.

Comment (lidavidm) [reply cpp/src/arrow/dataset/file_base.cc:144]: Ah good point. It is mostly just a nit as it's really a testing parameter that's unfortunately getting exposed in the public API.

This isn't a very strong precedent, but TableBatchReader handles batch_size by letting you set it after construction and that feels like an analogue of this.

Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc:144]: Once SyncScanner goes away we could probably change `InMemoryFragment::record_batches_` to `InMemoryFragment::record_batch_`.  This reflects the spirit of "getting rid of scan tasks" better anyways.

File: cpp/src/arrow/dataset/scanner.cc
Comment (lidavidm) [inline cpp/src/arrow/dataset/scanner.cc:247]: nit: all the other methods return Status (even if it's a bit pointless)

File: cpp/src/arrow/dataset/scanner.h
Comment (lidavidm) [inline cpp/src/arrow/dataset/scanner.h:172]: FWIW I had moved these helpers to the bottom of the file (outside the `@}`) so that they don't get picked up in the API docs as I'd basically consider them implementation details. Though at that point we may actually want them to be in one of the `_internal` headers.

File: cpp/src/arrow/dataset/scanner.h
Comment (lidavidm) [inline cpp/src/arrow/dataset/scanner.h:340]: Is this worth declaring in the header? We could keep it in the .cc file so that people don't instantiate it directly.

**Comment by lidavidm** [comment]
After this lands I can rebase and implement ScanBatchesAsync for IPC/Parquet and give that another test.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h:340]: Ah, that's a good idea.  I can move both the scanners to `.cc`.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h:172]: Oh shoot.  After the rebase I had two instances of the exact same method and this explains it.  I'll fix that up.

File: cpp/src/arrow/dataset/scanner_test.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner_test.cc]: I've absorbed `UseThreads` into the matrix.

File: cpp/src/arrow/dataset/file_base.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc]: I capitulated and removed the argument.  Your comment about it just being a testing parameter is accurate.  I created a test in `arrow-dataset-file-test` that does not rely on `InMemoryDataset` to test this logic here.  I might in the future add some tests to `ScannerTest` that set a limit on scan options batch size to get coverage of the multiple batches per fragment case.

File: cpp/src/arrow/dataset/file_base.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/file_base.cc]: Done.

File: cpp/src/arrow/dataset/scanner.cc
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.cc]: Fair enough.  I changed it to return a Status.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h]: Fixed.

Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h]: Well, for clarity's sake.  I just moved them back to where you had them.  I"ll leave internal headers for a future PR to keep this one at a reasonable size.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h]: Done.

File: cpp/src/arrow/util/iterator_test.cc
Comment (lidavidm) [inline cpp/src/arrow/util/iterator_test.cc:164]: If I'm not mistaken, FailsAt and AssertBufferIteratorMatch are now entirely unused?

File: cpp/src/arrow/dataset/scanner.h
Comment (lidavidm) [inline cpp/src/arrow/dataset/scanner.h:55]: I think moving the subclass definitions means you can also move this alias and get rid of the async_generator.h include.

**Comment by westonpace** [comment]
I rebased @lidavidm 's latest changes.  At this point I think I don't think there are anymore outstanding dataset PRs to rebase so I think this one is probably ready to merge if it passes review.

**Review by lidavidm — APPROVED** [review]
I think this is ready, just a couple more minor things.

File: cpp/src/arrow/util/iterator_test.cc
Comment (westonpace) [reply cpp/src/arrow/util/iterator_test.cc]: Correct.  I've removed them.

File: cpp/src/arrow/dataset/scanner.h
Comment (westonpace) [reply cpp/src/arrow/dataset/scanner.h]: I was able to move FragmentGenerator but `async_generator.h` was still needed for `Enumerated` which is a pity since `Enumerated` is small and self-contained.  Should I place it in its own file?

Comment (lidavidm) [reply cpp/src/arrow/dataset/scanner.h]: Ah, that's alright then.

**Comment by lidavidm** [comment]
I'll merge on green unless somebody else wants to take a look.

**Comment by lidavidm** [comment]
Integration build: usual issues
All MacOS builds: ARROW-12467 (need to account for LLVM 12)
Travis: I think this is a flake
AppVeyor: this is a flake, albeit a common one

**Pull request closed without merging**
